### ISSUE TITLE
Bug: sklearn.KernelDensity da log-densidades erróneas en cola profunda (afecta clasificador KDC)

### ISSUE THREAD

**Issue by capitantoto** [capitantoto/fermat#31]
## Resumen

Durante el desarrollo de la tesis se detectó que `sklearn.neighbors.KernelDensity` devuelve **log-densidades erróneas por cientos de nats** en régimen de cola profunda (query dentro del bounding box del entrenamiento, bandwidth chico). El bug afecta específicamente a la rama euclídea del clasificador `KDClassifier(metric="euclidean")` (i.e., `#kdc` en la notación de la tesis); `#fkdc`, `#fkn`, `#kn` y el resto de clasificadores están inafectados.

Un fix del lado de este repositorio ya existe en la rama `investigacion-v6` (commit [`0c9d795`](https://github.com/capitantoto/fermat/commit/0c9d795)); este issue lo documenta y propone un plan de integración.

## Reproductor mínimo

Standalone, sin dependencia de `fkdc`, con `numpy`/`scipy`/`sklearn`: [`scripts/reproduce_sklearn_kde_bug.py`](https://github.com/capitantoto/fermat/blob/investigacion-bug-sklearn-kde/scripts/reproduce_sklearn_kde_bug.py). Salida al ejecutarlo:

\`\`\`
Train: 100 puntos en R^3 (círculo de radio 1.5 en xy)
Query: 3 puntos dentro del bounding box del train
Bandwidth: h = 0.05
Distancia mínima query→train: [1.5 0.94 1.08]

--- Comparación (log-densidades) ---
      sklearn KDTree     sklearn leaf=N+1     logsumexp exacto
            -29.8347            -443.7883            -443.7883
           -174.3540            -174.3540            -174.3540
           -228.9246            -228.9246            -228.9246

Error máximo sklearn KDTree vs exacto: 4.140e+02
Error máximo sklearn leaf=N+1 vs exacto: 1.705e-13
\`\`\`

En el primer punto (el centro del círculo, distancia 1.5 al train más cercano) sklearn devuelve $-29.83$ donde el valor verdadero es $-443.79$: un error de **414 nats**.

## Mecanismo

Del análisis forense en `docs/sandbox/dim-mixta-forense.md`:

> \`sklearn/neighbors/_binary_tree.pxi\` mantiene cotas globales en log-espacio (\`global_log_min_bound\`, \`global_log_bound_spread\`) con cadenas de \`logaddexp\`/\`logsubexp\`. Cuando el punto de consulta cae dentro de la caja envolvente de la clase (\`dist_LB(raíz)=0\` ⇒ spread inicial ≈ log N) pero lejos de todos los puntos, la cancelación catastrófica de los \`logsubexp\` deja un residuo de orden eps·N que el recorrido nunca limpia, y la densidad devuelta queda anclada cerca de \`ln(eps) + log N ≈ −31\`.

Reproducido con sklearn 1.7.2, tanto \`kd_tree\` como \`ball_tree\`, breadth y depth first, con \`atol=rtol=0\`. Con \`leaf_size ≥ N\` (árbol de una sola hoja ⇒ suma exacta dentro de la hoja) el error cae a $5 \\cdot 10^{-13}$.

## Impacto en los resultados de la tesis

El bug afectó al pipeline v5 sobre el que corren los resultados de §5. De las 500 corridas de #kdc (20 datasets × 25 semillas), **417 son bit-a-bit idénticas** al recomputarlas con un KDE euclídeo exacto: el bug solo se activa cuando la CV explora anchos de banda chicos en régimen de cola profunda.

El cambio material se concentra en \`helices_0\`:

| Dataset | R² mediano #kdc (viejo) | R² mediano #kdc (corregido) | Δ mediana |
|---|---:|---:|---:|
| \`helices_0\` | 0.9199 | 0.9714 | **+0.0515** |
| \`espirales_lo\` | 0.8303 | 0.8321 | +0.0019 |
| \`eslabones_0\` | 0.9924 | 0.9912 | −0.0013 |
| 15 datasets restantes | idénticos | idénticos | 0 |

Consecuencia: la ventaja de \"casi 5 puntos porcentuales\" de #fkdc sobre #kdc en \`helices_0\` reportada en §5.1 (l.2018 de \`docs/tesis.typ\`) desaparece — era un artefacto del bug. La conclusión central (paridad-o-mejora al ampliar el espacio de hiperparámetros con la distancia de Fermat, con simbiosis de la regla R1SD) **se refuerza** con el fix: la mejora atribuible a la distancia de Fermat es menor de lo reportado, pero la mejora atribuible a la ampliación del espacio de hiperparámetros aumenta.

Ver \`docs/sandbox/erratum-v5-kdc.md\` (rama \`investigacion-v6\`) para los 14 claims específicos afectados con veredicto por línea.

## Fix propuesto (ya existe en \`investigacion-v6\`)

Commit [\`0c9d795\`](https://github.com/capitantoto/fermat/commit/0c9d795), \`fkdc/fermat.py\`:

1. Nueva clase \`DistanciaEuclidea\` con la interfaz mínima de \`SampleFermatDistance\` (usa \`euclidean.func\` sobre \`cdist\`).
2. Nueva clase \`EuclideanKDE(FermatKDE)\`: hereda \`score_samples\` (misma suma exacta de kernels sobre la matriz de distancias, idéntica normalización, idéntico clip).
3. En \`KDClassifier.fit\`, la rama \`metric=\"euclidean\"\` se rutea a \`EuclideanKDE\` en lugar de \`sklearn.KernelDensity\`.

Consecuencias:
- **#fkdc(α=1) ≡ #kdc bit-a-bit** por construcción (mismo camino de código).
- Elimina también una asimetría de clipping secundaria entre \`FermatKDE.score_samples\` y \`KDClassifier.predict_proba\` que impedía la equivalencia estricta.
- Costo de cómputo: O(n²) por fuerza bruta, aceptable para los tamaños de dataset de la tesis (n < 1000).

Alternativa mínima (no recomendada — deja viva la asimetría de clipping): \`KernelDensity(..., leaf_size=len(X)+1)\`. Verificada exacta a $5 \\cdot 10^{-13}$ en el reproductor.

## Plan de integración

- [ ] Rebasear \`investigacion-v6\` sobre \`main\` (o cherry-pick \`0c9d795\`) → nueva rama \`fix-kdc-euclidean\`.
- [ ] Portar el test \`docs/sandbox/test-kdc-fix.py\` a \`tests/\` como test suite regular del repo.
- [ ] Regenerar los CSVs de \`docs/data/\` afectados (los de \`helices_0-*\` principalmente) y actualizar §5.1 y §5.4 en \`docs/tesis.typ\` para reflejar los números post-fix. Decisión pendiente del autor: (a) integrar como corrección al cuerpo, (b) mantener el cuerpo tal cual y agregar nota al pie que remita al anexo IA (donde ya está registrado), o (c) publicar como corrigendum separado.
- [ ] Abrir PR contra \`main\`.

## Upstream

Estado actual de los tickets originales de scikit-learn:

- [**scikit-learn#27186**](https://github.com/scikit-learn/scikit-learn/issues/27186) — **abierto**, etiqueta \"Needs Investigation\", identifica el origen probable en \`node_log_bound_spreads\` (usa la cota máxima en vez del spread). Sin respuesta de mantenedores.
- [**scikit-learn#25799**](https://github.com/scikit-learn/scikit-learn/issues/25799) — **cerrado sin fix** (\"not planned\"), reporta el mismo síntoma con bandwidth chico.

Merece comentar en #27186 con el reproductor mínimo de este repo, y quizás abrir un PR upstream con el fix propuesto en #27186 (usar \`logsubexp\` para computar el spread real). El fix del lado de fkdc no depende de que upstream se resuelva; es prudente porque además garantiza la identidad estricta \`#fkdc(α=1) ≡ #kdc\`.

## Referencias

- Análisis forense: \`docs/sandbox/dim-mixta-forense.md\` (rama \`investigacion-v6\`)
- Erratum vs. tesis: \`docs/sandbox/erratum-v5-kdc.md\` (rama \`investigacion-v6\`)
- Fix existente: commit [\`0c9d795\`](https://github.com/capitantoto/fermat/commit/0c9d795) en \`investigacion-v6\`
- Reproductor standalone: \`scripts/reproduce_sklearn_kde_bug.py\` (esta rama)
- Test suite del fix: \`docs/sandbox/test-kdc-fix.py\` (rama \`investigacion-v6\`)
- Anexo IA de la tesis: §Anexo (l. 2311 de \`docs/tesis.typ\` en \`main\`) documenta el hallazgo como \"pendiente de validación independiente\"

🤖 Investigación y redacción asistidas por IA (Claude); reproductor verificado independientemente.

**Comment by capitantoto** [comment]
### Rehearsal PR (upstream): [capitantoto/scikit-learn#1](https://github.com/capitantoto/scikit-learn/pull/1)

Forkeé `scikit-learn/scikit-learn` a `capitantoto/scikit-learn` y abrí un PR **draft** contra el fork con:
- El fix real en `_binary_tree.pxi.tp` (populado `node_log_bound_spreads[i]` con `logsubexp(log_max, log_min)` en vez de solo `log_max`), directo del diagnóstico del reporter original de scikit-learn#27186.
- Un test de regresión parametrizado (`kd_tree` + `ball_tree`) en `sklearn/neighbors/tests/test_kde.py` que reproduce el bug con la misma fixture de círculos anidados que el standalone.
- Referencia cruzada al análisis forense del sandbox (`docs/sandbox/dim-mixta-forense.md` en rama `investigacion-v6`) y al reproductor standalone del repo (`scripts/reproduce_sklearn_kde_bug.py`).

**Pendientes antes de eventualmente moverlo upstream** (todos anotados en el body del PR draft):
1. CHANGELOG upstream de sklearn.
2. Benchmark del efecto del fix sobre la convergencia del recorrido (usar el spread real puede ajustar más los criterios de pruning y hacer más trabajo en casos benignos — no medido).
3. Compilar Cython y correr el test suite completo de sklearn localmente para verificar que ningún otro test se rompe (no lo hice — el fork no tiene CI habilitado y no crucé el paso de compilar Cython aquí).
4. Revisión editorial del wording del comentario inline.

Decisión pendiente del autor: si mover a upstream tal cual, si primero abrir un comentario en scikit-learn#27186 con el reproductor mínimo para pedir validación de mantenedores, o si integrar únicamente el fix del lado fkdc y dejar el upstream para más adelante.
